src/coreclr/vm/gcenv.ee.cpp

```diff
@@ -55,6 +55,27 @@ void GCToEEInterface::RestartEE(bool bFinishedGC)
 {
     WRAPPER_NO_CONTRACT;
 
+    // The GC may change allocation contexts during the GC using GcEnumAllocContexts() so we need to
```

Comment:
It is unfortunate that we have do a dedicated thread enumeration of the thread list just to clear the single field. Is there really no better way to do that?

cc @VSadov

Reply:
This can be noticeable on scenarios like asp.net benchmarks (relatively small heap, but lots of threads), but may not be a big deal overall next to the cost of a GC.

Perhaps should check that suspend reason was indeed `SUSPEND_FOR_GC`. 
I'd assume that other kinds of suspension like `SUSPEND_FOR_GC_PREP` will not be adjusting allocation budgets.

Reply:
> Is there really no better way to do that?

I chatted with Maoni and she was comfortable with us codifying as part of the current GC contract that the only legal modification the GC can make inside the enumeration is to set ptr = limit = 0. Once the EE is allowed to assume that additional information then we can change the enumeration code to look like this:

```
fn(alloc_context, param);
if(alloc_context->alloc_limit == 0)
{
    combined_limit = 0;
}
```

Concurrent threads enumerating will still create races, but the races will no longer matter. How does that sound?

Reply:
I just noticed your comment https://github.com/dotnet/runtime/pull/104849#discussion_r1677867717 @jkotas. We can update combined_limit inline inside GCEnumAllocContexts() or we can do it with a separate loop through all the threads afterwards.

Reply:
I've updated the PR with the new approach that zeros combined_limit inside the GCEnumAllocContexts() function and avoids doing a separate thread enumeration in RestartEE().

Reply:
> Concurrent threads enumerating will still create races,

It is also going to create cache contention.

It would be much easier to reason about an explicit contract where the GC can say that it is doing the final fixup of the allocation context.

Reply:
The plan for now is to do some perf testing with a large number of threads but @Maoni0 did not expect this to be an issue. She said that past GC performance investigations had not raised these thread enumerations playing a significant role.

Fwiw, the GC's existing code already creates similar contention here. N GC server threads run the thread enumeration in parallel and the callback both reads and potentially writes to fields that reside in the same cache line. https://github.com/dotnet/runtime/blob/main/src/coreclr/gc/gc.cpp#L7888-L7896. The likely case is that whatever thread dirties the cache line writing alloc_ptr=0 and alloc_limit=0 will be the same thread that writes combined_limit=0 a few CPU instructions later.

Reply:
@jkotas - For now the options I am aware of are either the RestartEE() separate thread enumeration or the inline updates on GcEnumAllocContexts(). I am not aware of any other options that don't require changes to the GC<->EE interface and @Maoni0 doesn't have time available to discuss/design/approve such changes. You've raised criticism on both choices, but I think the path forward requires either picking one of them or suggesting a new 3rd option which doesn't require GC interface changes to implement.

Reply:
I think we should be open to adding new methods to GC<->EE interface as necessary. GC<->EE interface was not designed as internal implementation details that is not set in stone, it is very similar to JIT<->EE interface. Changes in JIT<->EE interface is no-brainer that allows us to evolve the system as something that easy to reason about and efficient. I do not understand why changing the GC<->EE interface is such a big deal.

I think that the design of the new GC<->EE interface method should be pretty straightforward: Add new variant of `GcEnumAllocContexts` that the GC is going to use when it is repairing the contexts to let the EE side know that it needs to do its part. Am I missing something?

Reply:
We are effectively at 'disagree and commit'. @Maoni0 does not wish to invest in any GC<->EE interface changes right now as a matter of prioritization. @jkotas notes his disagreement, but we'll move forward with no GC<->EE change and the PR will stay as is. I will confirm its not causing a perf issue.

Reply:
I think that this change is regressing the GC/EE interface design. As discussed offline, I can live with it but I am not happy about it.